lib/srv/db/common/databaseobjectimportrule/import_rules.go

```diff
@@ -52,46 +57,122 @@ func ApplyDatabaseObjectImportRules(rules []*dbobjectimportrulev1.DatabaseObject
 		}
 	}
 
+	var objects []*dbobjectv1.DatabaseObject
+	errors := map[string]error{}
+
 	// anything to do?
 	if len(mappings) == 0 {
-		return nil
+		return objects, errors
 	}
 
-	var out []*dbobjectv1.DatabaseObject
-
 	// find all objects that match any of the rules
 	for _, obj := range objs {
-		var objClone *dbobjectv1.DatabaseObject
+		// prepare object clone
+		objClone := utils.CloneProtoMsg(obj)
+		if objClone.Metadata.Labels == nil {
+			objClone.Metadata.Labels = map[string]string{}
+		}
 
 		// apply each mapping in order.
+		matched := false
+		hadError := false
 		for _, mapping := range mappings {
-			// the matching is applied to the object spec; existing object labels does not matter
-			if !databaseObjectScopeMatch(mapping.GetScope(), obj.GetSpec()) {
-				continue
+			match, err := applyMappingToObject(mapping, objClone.GetSpec(), objClone.Metadata.Labels)
+			if err != nil {
+				errors[obj.GetMetadata().GetName()] = trace.Wrap(err)
+				hadError = true
+				break
 			}
-			if databaseObjectImportMatch(mapping.GetMatch(), obj.GetSpec()) {
-				if objClone == nil {
-					objClone = utils.CloneProtoMsg(obj)
-				}
-
-				// mapping applies additional labels
-				labels := objClone.Metadata.Labels
-				if labels == nil {
-					labels = map[string]string{}
-				}
-				for k, v := range mapping.AddLabels {
-					labels[k] = v
-				}
-				objClone.Metadata.Labels = labels
+			if match {
+				matched = true
 			}
 		}
 
-		if objClone != nil {
-			out = append(out, objClone)
+		if !hadError && matched {
+			objects = append(objects, objClone)
+		}
+	}
+
+	return objects, errors
+}
+
+// validateTemplate evaluates the template, checking for potential errors.
+func validateTemplate(template string) error {
+	_, err := evalTemplate(template, nil)
+	return trace.Wrap(err)
+}
+
+func evalTemplate(template string, spec *dbobjectv1.DatabaseObjectSpec) (string, error) {
```

Comment:
What's the difference between this impl of `evalTemplate` and calling `parse.NewTraitsTemplateExpression` + `expr.Interpolate`?

Reply:
One problem is that `parse.NewTraitsTemplateExpression` ultimately calls `newTraitsTemplateParser` which uses `typical.WithInvalidNamespaceHack()`. 

The other is that it hardcodes `external` and `internal` as the only two available variable names, instead of `obj` that I wanted to use here (because it is more natural).

Finally, the traits are built around `[]string` results, instead of `string`. Bypassing `TraitsTemplateExpression.Interpolate` allows us to avoid the singleton list awkwardness.

Reply:
I see. `NewTraitsTemplateExpression` has `Traits` in it so it's fine if it hardcodes `external` and `internal`. 

I am not a big fan of exposing some internals from `parse`. Do you think if we can generalize this implementation a little bit and keep it in `parse`?

Reply:
The underlying library here is `typical`, which is also our own code. `parse` sounds general, but is truly about traits, and I think we should avoid adding extra features there. There is also [`lib/expression`](https://github.com/gravitational/teleport.e/tree/master/lib/expression) library in `teleport.e`, but it is also based on `typical`.

The only thing used from `parse` is `parse.SplitExpression`, which is a newly exposed utility function. I'm strongly considering breaking this dependency completely - revert the changes to `parse` and just reimplement this function locally; it isn't complex and this would disentangle this unnecessary dependency.

Reply:
> The only thing used from parse is parse.SplitExpression, which is a newly exposed utility function. I'm strongly considering breaking this dependency completely - revert the changes to parse and just reimplement this function locally; it isn't complex and this would disentangle this unnecessary dependency.

Agreed.

Reply:
Done, PTAL.